
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/Users/Dyll/Documents/Education/VU_UVA/Internship/Epigenetics/Janssen_Group-UMCUtrecht/Main_Project")
```
Load the libraries
```{r}
library(dplyr)
library(broom)
library(knitr)
library(coin)
library(data.table)
library(ggpubr)
library(rstatix)
# library(ggplot2)
# library(paletteer)
# library(PairedData)
library(tidyverse)
``` 

Load other options
```{r}
# options(scipen = 1, digits = 20)
getOption("max.print")
```


Load the RNA metrics dataset and remove the useless columns. 

The data shows the samples as rows and descriptive characteristics, the Tissue Source Site (TSS) and the Histology grouping as columns.

```{r}

random_rows <- sample(1:11060, 500)

first <- fread("Data/RNA_Data/Full_RNA_metrics1.csv")

used_first <- first[random_rows,c("Sample", "Mean_set", "Mean_cntrl","TSS","Hist_group", "TSS_Abbrvs")]
setnames(used_first, old = "Mean_cntrl", new = "Mean_cntrl1")
used_first[, Mean_diff1 := (Mean_set - Mean_cntrl1)]

num_datasets <- 412

# Loop over the datasets
for (i in 2:num_datasets) {
  
  file_name <- paste("Data/RNA_Data/Full_RNA_metrics", i, ".csv", sep = "")
  

  # Read the CSV file into a data.table
  current_dataset <- fread(file_name)
  current_dataset[, Mean_diff := (Mean_set - Mean_cntrl)]

  new_name1 = paste('Mean_cntrl', i, sep = "")
  new_name2 = paste('Mean_diff', i, sep = "")

  rename_map <- c("Mean_cntrl" = new_name1, "Mean_diff" = new_name2)

  # Rename columns in 'current_dataset' using the named vector
  setnames(current_dataset, old = names(rename_map), new = rename_map)

  # Select columns with their original names from 'current_dataset' and rename them
  selected_columns <- current_dataset[random_rows, c("Sample", new_name1, new_name2), with = FALSE]

  # Perform a merge based on Sample ID
  used_first <- merge(used_first, selected_columns, by = "Sample", all = FALSE, suffixes = c("", i))

}
``` 

Part I. Inter-genomic expression analysis 


QUESTION 1 


A) Create violin plots of the differences of the average log2 expression to assess the type of paired test to perform on the the pairwise groups .

```{r}


violin_graph_long_data <- used_first %>% 
  dplyr::select(sample(starts_with("Mean_dif"), 7)) %>% 
  pivot_longer(cols = everything(),
               names_to = "Variable",
               values_to = "Value") 



ggplot(violin_graph_long_data, aes(y = Value, x = Variable)) +
  geom_violin(trim = FALSE)+
  geom_boxplot(varwidth = T, alpha  = 0.25, outlier.colour = "red")+
  labs(x = "Unique SoI / Control Pairings ", y = "Δ Log2 Expression",
       title = "Distribution of the Set-of-Interest to Control Set \n Differential Log2 Expression Across Various Pairings") +
  theme(
    legend.position = "none",
    axis.title = element_text(size = 13, color = "black", face = "bold"),
    axis.text = element_text(color = "black", size = 10),
    axis.text.x = element_blank(),
    panel.background = element_rect(fill = "#EDEDED"),
    legend.text = element_blank(),
    plot.title = element_text(hjust = 0.6, size = 15),
    panel.border = element_rect(fill = "transparent", color = 1, linewidth = 1),
    panel.grid.major.y = element_line(color = "black", linewidth = 0.1, linetype = 2),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_line(color = "black", linewidth = 0.1, linetype = 2),
    plot.title.position = "plot"
  )


``` 
Interpretation: Plots clearly show distinct pairings which violate a normal distribution. This follows that a non-parametric test of paired differences between the means can be performed for all pairings. This method is more stringent but more efficient than evaluating each distribution individually to decide on whether to apply a parametric or non-parametric test to the distributions. 
The data is however not symmetrical around the median and the paired Wilcoxon test has this as an assumption which will come into play for the statistical comparisons between the SoI and control sets. Should not be an issue when comparing amongst each other though as the skew is in the same direction for all. 


B) Use a Friedman's test (non-parametric version of the RMA, which is a paired t-test for more than 2 groups) to test whether the different control groups differ significantly in their distributions.


```{r}
Fried_test_data <- used_first %>%
  gather(key = "Pairing", value = "Differential", starts_with("Mean_cntrl")) %>%
  mutate(Sample = as.factor(Sample), Pairing = as.factor(Pairing)) %>% 
  select(Sample, Pairing, Differential)

friedman_result <- coin::friedman_test(Differential ~ Pairing | Sample, data = Fried_test_data)

```
Result: The Friedman Test shows that at least one of the pairings demonstrates a significantly different distribution. However, performing pairwise post hoc tests between all pairs of groups may not be practical due to the combinatorial explosion of comparisons



C) Construct a contrast analysis to assess each of the pairings against the general trend

```{r}
Paired_wilc_table <- used_first %>% 
  dplyr::select(starts_with("Mean_diff")) %>%
  mutate_all(as.numeric) 
   
Contrast_table <- data.frame(Set = character(0), P_Value = numeric(0), Adj_P_Value = numeric(0), Significant = character(0))

# Loop over each control sets
for (i in 1:num_datasets) {
  
  set = paste('Mean_diff', i, sep = "")
  
   Global_table <- Paired_wilc_table %>%
    mutate(global = rowMeans(select(., -set), na.rm = TRUE)) %>%
    select(set, global)
   
   wlcx_paired_res = wilcox.test(Global_table[,get(set)], Global_table$global, paired = TRUE)
   Corrected_Pval <- wlcx_paired_res$p.value * num_datasets
   Significant_value <- if(Corrected_Pval < 0.05) "TRUE" else "FALSE"
   
   Contrast_table <- rbind(Contrast_table, data.frame(Set = set, P_Value = wlcx_paired_res$p.value, Adj_P_Value = Corrected_Pval, Significant = Significant_value)) %>% arrange(desc(Adj_P_Value))
}


```
Result: The Contrast analysis shows that the expression pattern of the control sets seem to mostly differ significantly compared to their average with only 25/412 showing an pvalue > 0.05 for the paired Wilcoxon Signed Rank Test. Cannot therefore assume that all the control sets behave in the same general fashion. 


C) Create plots of the differences of the average log2 expression across all pairings between the SoI and controls sets in order to visualise the similarities fo the behaviour between the controls and the SoI. 
Seeing as the Friedman test came back indicating that at least one of the controls shows a different expression distribution compared to the others, cannot use the test as evidence to dismiss all controls but one. Instead we can visualise the general distribution across all controls to determine if they act in a similar fashion. 

```{r}

graph_long_data <- used_first %>% 
  dplyr::select(starts_with("Mean_diff")) %>% 
  pivot_longer(cols = everything(),
               names_to = "Variable",
               values_to = "Value") 



ggplot(graph_long_data, aes(y = Value, x = Variable)) +
  geom_boxplot(varwidth = T, alpha  = 0.25, outlier.colour = "#EDEDED")+
    coord_cartesian(ylim = c(0, 4)) +
  scale_y_continuous() +
  labs(x = "Unique SoI / Control Pairings ", y = "Δ Log2 Expression",
       title = "Distribution of the Set-of-Interest to Control Set \n Differential Log2 Expression Across All Pairings") +
  theme(
    plot.title = element_text(hjust = 0.6, size = 15),
    plot.title.position = "plot",
    axis.title = element_text(size = 13, color = "black", face = "bold"),
    axis.text = element_text(color = "black", size = 10),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.text = element_blank(),
    legend.position = "none",
    panel.border = element_rect(fill = "transparent", color = 1, linewidth = 1),
    panel.grid.major.y = element_line(color = "black", linewidth = 0.1, linetype = 2),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_line(color = "black", linewidth = 0.1, linetype = 2),
    panel.background = element_rect(fill = "#EDEDED")
)
``` 
Interpretation: Despite that fact that the Friedman Test resulted in a significant result (indicating a significant difference in differential expression with at least one of the control groups), the plot above clearly demonstrates that the differential expression pattern between every control set and the Set-of-Interest is bound in a consistent range centered between log2(2) and log2(3). Indicating the expression within the SoI is always increased compared 



D) Calculate the significance of the difference between the mean expression of the control-sets and the Set-of-Interest using Wilcoxon Signed-Rank Tests for each control/SoI pair followed by a correction for multiple testing. 

```{r}
# Loop over each of the control datasets in the folder.
# For each control-set perform a Two-Sample wilcoxon Signed-Rank test with the SoI.
# Save the pvalue for each test in a dataframe.

paired_tests_list <- list()
for (i in 1:num_datasets) {
  ctrl <- paste("Mean_cntrl", i, sep ="")
  wlcx_paired_res = wilcox.test(used_first$Mean_set, used_first[,get(ctrl)], paired = TRUE)
  wlcx_paired_res_pval <- wlcx_paired_res$p.value
  formatted_pval <- sprintf("%.10f", wlcx_paired_res_pval)
  pair <- paste("Pair", i, sep ="_")
  pair_frame = data.frame(Pairing = pair, Pvalue = as.numeric(formatted_pval))
  paired_tests_list[[i]] <- pair_frame
}

# Correct the P-value due to multiple testing by using the Bonferoni method (more stringent than BH).

full_paired <- do.call(rbind, paired_tests_list) %>% arrange(desc(Pvalue))
full_paired$Corrected_Pval <- full_paired$Pvalue * num_datasets
```
Interpretation: the p-value shows a significant difference between the means of the paired sets. 




E) Demonstrate difference by showing the individual expression distributions of the SoI
  and some (3) of the control sets

```{r}
# Select 3 control sets at random to compare their distributions with that of the SoI.

controls <- c("Mean_set")
for (i in sample(1:num_datasets, 3)) {
  controls <- c(controls, paste("Mean_cntrl", i, sep =""))
}

# Pivot the data into a form that can be used by boxplot.

long_data <- used_first %>% 
  dplyr::select(all_of(controls)) %>% 
  pivot_longer(cols = everything(),
               names_to = "Variable",
               values_to = "Value")


ggplot(long_data, aes( y = Value, x = Variable)) +
  geom_boxplot(outlier.colour = 2) +
  stat_boxplot(geom = "errorbar") +
  theme_minimal() +
  labs(title = "Distribution of the Mean Copy Number in the Set-of-Interest vs the Control-Set",
       y = "Log2 Expression") +
  # scale_x_discrete(label = c("Control Set", "Set of Interest")) +
  # scale_fill_hue(labels = c("Control Set", "Set of Interest"))+
  theme(
        legend.position = "None",
        axis.title.y = element_text(size= 13),
        plot.subtitle = element_text(hjust= 0.5),
        axis.title.x = element_blank(),
        axis.text.x = element_text(size = 11),
        plot.title = element_text(hjust = 0.5,
                                  size = 14),
        panel.background = element_rect(fill = "#EDEDED"),
        panel.grid.major.y = element_line(color = "black", linewidth = 0.25, linetype = 2),
        panel.border = element_rect(fill = "transparent",
                                    color = 1,
                                    linewidth = 2)) 

```
Interpretation: the p-value and plots demonstrate that the expression pattern of the control sets are significantly different from that of the SoI.

Impact on analyses: future test can now use a single control set to test against as it has been shown that the controls do not differ significantly from one another and all share a similar difference with the SoI.








QUESTION 2



Data using only a single control group
```{r}
main_control <- controls[2]

single_control <- used_first %>%
  dplyr::select(all_of(c("Sample", "Mean_set", main_control, "TSS", "Hist_group", "TSS_Abbrvs"))) %>% 
  mutate(Dif = Mean_set - .data[[as.character(main_control)]], 
         Sample = factor(Sample, levels = Sample[order(TSS_Abbrvs)]))


```



A) Create a barplot of the samples vs their mean differences in the mean log2 RNA values coloured by TSS.

```{r}


ggplot(single_control, aes(x = Sample, y = Dif, fill = TSS_Abbrvs)) + 
  geom_col() +
  theme_minimal() +
  labs(y = "Δ Mean Log2 Expression",
       x = "Samples",
       title = "Expression Differential by Sample ")+
  scale_y_continuous(breaks = c(0, 1, 2, 3))+
  theme(
    axis.text.x = element_blank(),
    axis.title.y = element_text(size = 14), 
    axis.text.y = element_text(size = 11),
    axis.title.x = element_text(size = 14), 
    legend.position = "bottom",
    legend.text = element_text(size= 9),
    legend.title =element_text(hjust = 0.5,
                               size = 13),
    plot.title = element_text(hjust = 0.4,
                              size = 18),
    plot.title.position = "plot",
    panel.background = element_rect(fill = "#EDEDED"),
    panel.grid.major.y = element_line(color = "black", linewidth  = 0.5, linetype = 2),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    axis.line = element_line(linewidth = 0.35, linetype = "solid",
                             colour = "black"),
    panel.border = element_rect(fill = "transparent",
                                color = 1,
                                linewidth = 1)) +
  guides(fill = guide_legend(ncol = 8, 
                             title = "Source Tissue Type",
                             title.position = "top")) 

``` 
Interpretation: The positive ∆ expression between the SoI and the control set is maintained across all cancer types


B) Create a barplot of the samples vs their differences in the mean log2 RNA values coloured by Hist

```{r}

single_control_reorder <- used_first %>%
  dplyr::select(all_of(c("Sample", "Mean_set", main_control, "TSS", "Hist_group", "TSS_Abbrvs"))) %>% 
  mutate(Dif = Mean_set - .data[[as.character(main_control)]], 
         Sample = factor(Sample, levels = Sample[order(Hist_group)]))

ggplot(single_control_reorder, aes(x = Sample, y = Dif, fill = Hist_group)) + 
  geom_col() +
  theme_minimal() +
  labs(y = "Δ Mean Log2 Expression",
       x = "Samples",
       title = "Expression Differential by Sample ")+
  scale_y_continuous(breaks = c(0, 1, 2, 3))+
  theme(
    axis.text.x = element_blank(),
    axis.title.y = element_text(size = 14), 
    axis.text.y = element_text(size = 11),
    axis.title.x = element_text(size = 14), 
    legend.position = "bottom",
    legend.text = element_text(size= 11),
    legend.title =element_text(hjust = 0.5,
                               size = 13),
    plot.title = element_text(hjust = 0.4,
                              size = 18),
    plot.title.position = "plot",
    panel.background = element_rect(fill = "#EDEDED"),
    panel.grid.major.y = element_line(color = "black", size = 0.5, linetype = 2),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.x = element_blank(),
    axis.line = element_line(size = 0.35, linetype = "solid",
                             colour = "black"),
    panel.border = element_rect(fill = "transparent",
                                color = 1,
                                linewidth = 1)) +
  guides(fill = guide_legend(ncol = 5, 
                             title = "Histology Grouping",
                             title.position = "top"))

``` 




C) Assess the distribution of the data for the control set that is selected



```{r}
ggplot(single_control, aes(x = Dif)) +
  geom_histogram(aes(y = after_stat(density)),
                 colour = 1,
                 fill = "white",
                 bins = 50) +
  geom_density(lwd = 1,
               colour = 1) +
  labs(y = "Density", x = "Δ Log2 Expression",
       title = "Distribution of the Set-of-Interest to Control Set \n Differential Log2 Expression Across Samples") +
  theme(
    axis.title = element_text(size = 13, color = "black", face = "bold"),
    axis.text = element_text(color = "black", size = 10),
    panel.background = element_rect(fill = "#EDEDED"),
    plot.title = element_text(hjust = 0.6, size = 15),
    panel.border = element_rect(fill = "transparent", color = 1, linewidth = 1),
    panel.grid.major.y = element_line(color = "black", linewidth = 0.25, linetype = 1),
    panel.grid.major.x = element_line(color = "black", linewidth = 0.25, linetype = 1),
    panel.grid.minor.x = element_line(color = "black", linewidth = 0.1, linetype = 1),
    panel.grid.minor.y = element_line(color = "black", linewidth = 0.1, linetype = 1),
    plot.title.position = "plot"
  )

ggplot(single_control, aes(sample= Dif)) +
  stat_qq() + 
  stat_qq_line(color= "red") +
  labs(title = "Normal QQ Plot for Analysis of the Average Set-of-Interest to \n Control Set Differential Mean Expression in Cancer Samples",
       y = "Sample Quantiles",
       x = "Theoretical Quantiles") +
  theme(
    axis.text.x = element_text(size= 8),
    axis.title.x = element_text(size= 13),
    axis.title.y = element_text(size= 13),
    legend.title =element_text(hjust = 0.5,
                               size = 11),
    plot.title = element_text(hjust = 0.5,
                              size = 14),
    panel.grid.major = element_line(color = "black",
                                    linewidth = 0.5,
                                    linetype = 1),
    panel.grid.minor = element_line(color = "black",
                                    linewidth = 0.2,
                                    linetype = 1),
    panel.background = element_rect(fill = "#EDEDED"),
    panel.border = element_rect(fill = "transparent",
                                color = 1,
                                linewidth = 2)
  ) 
```















Create the functions that are going to perform the paired t-test for each cancer type on the mean and the median stats.

Function creates a df that calculates the mean difference over the provided set and applies a paired t-test returning mean difference as well as the p-value.

```{r}
perform_mean_test <- function(data) {
  test_result <- t.test(data$Mean_set, data$main_control, paired = TRUE)
  return(data.frame(mean_diff = mean(data$Mean_set - data$main_control),
                    p_value = test_result$p.value))
}

``` 


B) Apply the "perform_mean_test" function separately on each TSS group.


```{r}
mean_results <- data %>%
  group_by(TSS, Hist_group) %>%
  filter(n() > 1 & !is.na(Mean_set) & !is.na(Mean_cntrl)) %>%
  do(perform_mean_test(.))

``` 

C) Correct the p-values due to mutliple testing using the Benjamini-Hochberg method for FDR.

```{r}
results_mean <- mean_results %>%
  mutate(p_adjusted = p.adjust(p_value, method = "BH")) 
``` 

D) Replace p-values with a value of 0 with a very small number to avoid downstream errors that would occur when applying log transformations.


```{r}
min_nonzero_p <- min(results_mean$p_adjusted[results_mean$p_adjusted > 0])
results_mean$p_value_adjusted <- ifelse(results_mean$p_adjusted == 0, min_nonzero_p, results_mean$p_adjusted)

``` 

E) Apply a -log10 transformation to the p-values for readability on graphs.

```{r}
results_mean$neg_log_pvalue <- -log10(results_mean$p_value_adjusted)

``` 

F) Sort by Histology group for readability/interpretation of graph and legend.

```{r}
results_mean_complete <- results_mean %>% arrange(Hist_group) 
unique_tss <- unique(results_mean_complete$TSS)
results_mean_complete$TSS <- factor(results_mean_complete$TSS, levels = unique_tss)

``` 

G) Add significance threshold. 

```{r}
threshold <- -log10(0.01)
``` 

H) Create plot of each TSS group, based on mean differential expression vs. p-value.

```{r}
ggplot(results_mean_complete, aes(x = mean_diff, y = neg_log_pvalue, fill = TSS)) +
  geom_point(size = 5, shape = 21, stroke = 1) +
  geom_hline(yintercept = threshold, linetype = "dashed", color = "red") +
  geom_text(aes(x = 2.9, y = threshold, label = ""), vjust = -0.5, hjust = -1.5, color = "red") +
  theme_minimal() +
  coord_cartesian(xlim = c(2.15, 3.2)) +
  labs(title = "Comparative Analysis of Average Differential Expression in Set-of-Interest vs Control Set", 
       x = "Average Differential Log2 Expression", 
       y = "Adjusted P-Value (-log10)") +
  theme(legend.position = "bottom",
        legend.text = element_text(size= 9),
        axis.title = element_text(size = 14),
        legend.title =element_text(hjust = 0.5,
                                   size = 10),
        plot.title = element_text(hjust = 0.5,
                                  size = 16),
        panel.grid.major = element_line(color = "black",
                                        linewidth = 0.5,
                                        linetype = 1),
        panel.grid.minor = element_line(color = "black",
                                        linewidth = 0.2,
                                        linetype = 1),
        panel.background = element_rect(fill = "#EDEDED"),
        panel.border = element_rect(fill = "transparent",
                                    color = 1,
                                    linewidth = 2)) +
  scale_color_discrete(name = "Histology Group")

``` 



Part II. Intra-Gene-set & inter-cancer type expression analysis

Reorder the data based on the histology group for readability and interpretation of graph and legend.

```{r}
data_complete <- data %>% arrange(Hist_group) 
unique_Abbrvs <- unique(data_complete$TSS_Abbrvs)
data_complete$TSS_Abbrvs <- factor(data_complete$TSS_Abbrvs, levels = unique_Abbrvs)

``` 

Q1. 
A) Boxplot showing the various distributions of RNA for each cancer type.

```{r}
ggplot(data_complete, aes(x = TSS_Abbrvs, y = Mean_set, fill = Hist_group)) + 
  geom_boxplot() +
  stat_boxplot(geom = "errorbar") +
  theme_minimal() +
  labs(title = "Average Gene-set Expression by Source Tissue Type",
       y = "Log2 Expression",
       fill = "Histology Group") +
  theme(legend.position = "right",
        legend.text = element_text(size= 7),
        axis.text.x = element_text(angle= 90, 
                                   size= 8),
        axis.title.x = element_blank(),
        legend.title =element_text(hjust = 0.5,
                                   size = 11),
        plot.title = element_text(hjust = 0.5,
                                  size = 14),
        panel.grid.major = element_line(color = "black",
                                        linewidth = 0.5,
                                        linetype = 1),
        panel.grid.minor = element_line(color = "black",
                                        linewidth = 0.2,
                                        linetype = 1),
        panel.background = element_rect(fill = "#EDEDED"),
        panel.border = element_rect(fill = "transparent",
                                    color = 1,
                                    linewidth = 2)) 

``` 

B) For the ANOVA test between the groups means, need to test for: Normality of data within each group

```{r}
ggplot(data_complete, aes(x = Mean_set, fill = TSS_Abbrvs)) + 
  geom_histogram(alpha = 0.3, position = "identity") +
  theme_minimal() +
  labs(title = "Histogram of Gene-set Expression by Source Tissue Type",
       y = "Log2 Expression",
       fill = "TSS") +
  theme(legend.position = "right",
        legend.text = element_text(size= 7),
        axis.text.x = element_text(size= 8),
        legend.title =element_text(hjust = 0.5,
                                   size = 11),
        plot.title = element_text(hjust = 0.5,
                                  size = 14),
        panel.grid.major = element_line(color = "black",
                                        linewidth = 0.5,
                                        linetype = 1),
        panel.grid.minor = element_line(color = "black",
                                        linewidth = 0.2,
                                        linetype = 1),
        panel.background = element_rect(fill = "#EDEDED"),
        panel.border = element_rect(fill = "transparent",
                                    color = 1,
                                    linewidth = 2)) 
```

```{r}
ggplot(data_complete, aes(sample= Mean_set, color = TSS)) +
  stat_qq() + 
  stat_qq_line() +
  facet_wrap(~TSS) + 
  theme(
    legend.position = "None",
    axis.text.x = element_text(size= 10),
    plot.title = element_text(hjust = 0.5,
                              size = 19),
    panel.grid.major = element_line(color = "black",
                                    linewidth = 0.5,
                                    linetype = 1),
    panel.grid.minor = element_line(color = "black",
                                    linewidth = 0.2,
                                    linetype = 1),
    panel.background = element_rect(fill = "#EDEDED"),
    panel.border = element_rect(fill = "transparent",
                                color = 1,
                                linewidth = 2)
  ) 
```

C) For the ANOVA test between the groups means, need to test for: Homoscedasticity of data within each group 




